pkgs/applications/networking/instant-messengers/zulip/default.nix

```diff
@@ -1,36 +1,70 @@
 { lib
-, fetchurl
-, appimageTools
+, fetchFromGitHub
+, buildNpmPackage
+, electron
+, makeDesktopItem
+, copyDesktopItems
 }:
 
-let
+buildNpmPackage rec {
   pname = "zulip";
   version = "5.10.3";
 
-  src = fetchurl {
-    url = "https://github.com/zulip/zulip-desktop/releases/download/v${version}/Zulip-${version}-x86_64.AppImage";
-    hash = "sha256-AnaW/zH2Vng8lpzv6LHlzCUnNWJoLpsSpmD0iZfteFg=";
-    name="${pname}-${version}.AppImage";
+  src = fetchFromGitHub {
+    owner = "zulip";
+    repo = "zulip-desktop";
+    rev = "v${version}";
+    hash = "sha256-H9RiL/EHGBSjsacJj8VR8weSajmkcrD0S0gXvWkMkD0=";
   };
 
-  appimageContents = appimageTools.extractType2 {
-    inherit pname version src;
-  };
+  npmDepsHash = "sha256-4+JPjdgUF1dZFAAjxhVQJZZfD7SJZkM/O+xTuy2mmWs=";
+
+  ELECTRON_SKIP_BINARY_DOWNLOAD = 1;
+
+  nativeBuildInputs = [
+    copyDesktopItems
+  ];
+
+  dontNpmBuild = true;
+  buildPhase = ''
+    runHook preBuild
+
+    npm run pack -- \
+      -c.electronDist=${electron}/libexec/electron \
+      -c.electronVersion=${electron.version}
 
-in appimageTools.wrapType2 {
-  inherit pname version src;
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out/share/lib/zulip"
+    cp -r dist/*-unpacked/{locales,resources{,.pak}} "$out/share/lib/zulip"
```

Comment:
You haven’t provided any way for Zulip to find `locales` and `resources.pak` at that path. It doesn’t need them though—those are from Electron and already used from `${electron.unwrapped}/libexec/electron`.

Reply:
! I didn't realize that. I followed the same pattern as some other electron packages (if you grep for `locales,resources` you get 17 matches). Does this mean that just copying the app.asar file would be enough?

Reply:
In general, both `app.asar` and `app.asar.unpacked` are necessary (when they exist). For the specific case of Zulip, at this time, `app.asar.unpacked` only has the `fs-xattr` module which is only loaded on macOS. But we probably want to keep the whole `resources` directory to avoid surprises in the future.